cgen error, when example is compiled with -autofree

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.6 c689f80.f91f4d9
OS: linux, Ubuntu 20.04.6 LTS
Processor: 2 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz

getwd: /space/v/oo
vexe: /space/v/oo/v
vexe mtime: 2024-07-12 07:05:42

vroot: OK, value: /space/v/oo
VMODULES: OK, value: /home/delian/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.45.2
Git vroot status: weekly.2024.28-7-gf91f4d94
.git/config present: true

CC version: cc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
thirdparty/tcc status: thirdparty-linux-amd64 40e5cbb5

```

**What did you do?**
`v -g -o vdbg cmd/v && ./vdbg -autofree a.v`
```v
struct FlagData {
    raw        string  @[required]
    field_name string  @[required]
	delimiter  string
	name       string
	arg        ?string
	pos        int
	repeats    int = 1
}

arg := 'xyz'
d := FlagData{
	raw: 'abc'
	field_name: 'def'
	delimiter: ' '
	name: 'abc'
	arg: arg
	pos: 1
	repeats: 123
}
println(d)

```

**What did you expect to see?**

A compiled program, even when -autofree is used

**What did you see instead?**
```
==================
/tmp/v_1000/a.01J2JWAT8WFKSWX0531Y54BYZ3.tmp.c:13128: error: cannot convert 'struct string' to 'unsigned char'
...
==================
(Use `v -cg` to print the entire error message)

```

The generated C code in question:
```c
    main__FlagData d = ((main__FlagData){
        .raw = /*clone1*/_SLIT("abc"),
        .field_name = /*clone1*/_SLIT("def"),
        .delimiter = /*clone1*/_SLIT(" "),
        .name = /*clone1*/_SLIT("abc"),
        .arg = /*clone1*/ string_clone_static(arg),
        .pos = 1,
        .repeats = 123,
    });
```

Note, how for the `.arg` field, which is of type `?string`, the compiler generated just `string_clone_static(arg)` , which is wrong, since string_clone_static returns a string, not `?string` .

Without -autofree, the above is:
```c
        _option_string _t1;
        _option_ok(&(string[]) { arg }, (_option*)(&_t1), sizeof(string));
        main__FlagData d = ((main__FlagData){
        .raw = _SLIT("abc"),
        .field_name = _SLIT("def"),
        .delimiter = _SLIT(" "),
        .name = _SLIT("abc"),
        .arg = _t1,
        .pos = 1,
        .repeats = 123,
    });
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -640,7 +640,7 @@ fn (mut g Gen) struct_init_field(sfield ast.StructInitField, language ast.Langua
 	mut cloned := false
 	if g.is_autofree && !sfield.typ.is_ptr() && field_type_sym.kind in [.array, .string] {
 		g.write('/*clone1*/')
-		if g.gen_clone_assignment(sfield.typ, sfield.expr, sfield.typ, false) {
+		if g.gen_clone_assignment(sfield.expected_type, sfield.expr, sfield.typ, false) {
 			cloned = true
 		}
 	}

--- a/vlib/v/gen/c/testdata/assigning_and_printing_struct_with_optional_field.out
+++ b/vlib/v/gen/c/testdata/assigning_and_printing_struct_with_optional_field.out
@@ -1,0 +1,5 @@
+autofree is on
+SomeStruct{
+    val: 0
+    arg: Option('xyz')
+}

--- a/vlib/v/gen/c/testdata/assigning_and_printing_struct_with_optional_field.vv
+++ b/vlib/v/gen/c/testdata/assigning_and_printing_struct_with_optional_field.vv
@@ -1,0 +1,14 @@
+// vtest vflags: -autofree
+struct SomeStruct {
+	val int
+	arg ?string
+}
+
+x := 'xyz'
+s := SomeStruct{
+	arg: x
+}
+$if autofree {
+	println('autofree is on')
+}
+println(s)

